### PR TITLE
Chainable content builder

### PR DESCRIPTION
**Pull request by allait**
### Change ContentLoader.filter to return a new object

* Replaces YAMLLoader with ContentLoader. ContentLoader instances should
  be created for each manifest/section file on app startup.
* Moved .get_question method to ContentLoader. Return a copy of the
  question dictionary.
* ContentLoader.get_builder() will return a new builder instance with a full list of sections.
* ContentBuilder.filter now returns a new ContentBuilder instance, so filter calls can be chained:

      content_builder.filter({'lot': "SaaS"}).filter({'lot': "PaaS"})

  would return questions and sections relevant for both SaaS and PaaS.

Example use of the new ContentLoader and ContentBuilder classes:

`app/__init__.py`:

    sections_content = ContentLoader('sections_order.yml', 'app/content/g6/')

Creating a filtered ContentBuilder in a view method:

    content = section_content.get_builder().filter({'lot': "SaaS"})

Getting individual question data:

    content = section_content.get_question("questionId")

### 	Make ContentBuilder an iterator
 
Instead of `for section in content_builder.sections` it's now possible
to do `for section in content_builder`

Paired with @quis

### CODE REVIEW

File: dmutils/formats.py [hunk 1]
```diff
@@ -1,2 +1,39 @@
 DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
 DATE_FORMAT = "%Y-%m-%d"
+
+LOTS = [
+    {
+        'lot': 'saas',
+        'lot_case': 'SaaS',
+        'label': u'Software as a Service',
```
Comment (TheDoubleK) [inline dmutils/formats.py]: Why do labels need the 'u' modifier but others don't?

File: dmutils/content_loader.py [hunk 2]
```diff
@@ -111,15 +75,58 @@ def _question_should_be_shown(self, dependencies, service_data):
...
+        question_content["id"] = self._make_question_id(question)
+
+        return question_content
+
+    def _make_id(self, name):
+        return inflection.underscore(
+            re.sub("\s", "_", name)
+        )
+
+    def _make_question_id(self, question):
```
Comment (robyoung) [inline dmutils/content_loader.py]: This could be a staticmethod

File: dmutils/content_loader.py [hunk 3]
```diff
@@ -111,15 +75,58 @@ def _question_should_be_shown(self, dependencies, service_data):
...
+
+    def _load_question(self, question):
+        question_content = read_yaml(
+            self._directory + question + ".yml"
+        )
+        question_content["id"] = self._make_question_id(question)
+
+        return question_content
+
+    def _make_id(self, name):
```
Comment (robyoung) [inline dmutils/content_loader.py]: This could be a staticmethod

File: dmutils/formats.py [hunk 4]
```diff
@@ -1,2 +1,39 @@
...
+    {
+        'lot': 'scs',
+        'lot_case': 'SCS',
+        'label': u'Specialist Cloud Services',
+    },
+]
+
+
+def lot_to_lot_case(lot_to_check):
+    lot_i_found = [lot for lot in LOTS if lot['lot'] == lot_to_check]
```
Comment (TheDoubleK) [inline dmutils/formats.py]: This line isn't hard to parse at all.

File: dmutils/content_loader.py [hunk 5]
```diff
@@ -111,15 +75,58 @@ def _question_should_be_shown(self, dependencies, service_data):
...
+
+    def _populate_section(self, section):
+        section["id"] = self._make_id(section["name"])
+        section["questions"] = [
+            self.get_question(q) for q in section["questions"]
+        ]
+
+        return section
+
+    def _load_question(self, question):
```
Comment (robyoung) [inline dmutils/content_loader.py]: If directory were passed in this could be a staticmethod.

File: dmutils/content_loader.py [hunk 6]
```diff
@@ -111,15 +75,58 @@ def _question_should_be_shown(self, dependencies, service_data):
...
+    def __init__(self, manifest, content_directory):
         self._cache = {}
+        self._directory = content_directory
+
+        manifest_sections = read_yaml(manifest)
+
+        self._questions = {
+            q: self._load_question(q)
+            for section in manifest_sections
+            for q in section["questions"]
```
Comment (robyoung) [inline dmutils/content_loader.py]: Nested comprehensions make my brain hurt. This one isn't so bad because the variable names make it fairly obvious but I did have to stop and have a think about it.

File: dmutils/formats.py (hunk 1 above)
Comment (allait) [inline dmutils/formats.py]: Labels are the only ones that end up in the template and I think that caused unicode errors on py3, but I don't remember what the exact issue was.

File: dmutils/formats.py [hunk 7]
```diff
@@ -1,2 +1,39 @@
...
+    if lot_i_found:
+        return lot_i_found[0]['lot_case']
+    return None
+
+
+def get_label_for_lot_param(lot_to_check):
+    lot_i_found = [lot for lot in LOTS if lot['lot'] == lot_to_check]
```
Comment (robyoung) [inline dmutils/formats.py]: It's also duplicated here.

File: tests/test_content_loader.py [hunk 8]
```diff
@@ -17,354 +17,255 @@
...
-        return yaml.load(mocked_content[yaml_file])
-    return read
-
-
-@mock.patch("dmutils.content_loader.YAMLLoader.read")
 class TestContentBuilder(unittest.TestCase):
+    def test_content_builder_init(self):
+        content = ContentBuilder([])
+
+        self.assertEqual(content.sections, [])
```
Comment (robyoung) [inline tests/test_content_loader.py]: Elsewhere in dmutils we're using py.test. Is it worth replacing these with asserts?
